compendium/DeclarativeServices/test/SharedLibraryExceptionTest.cpp

```diff
@@ -0,0 +1,111 @@
+/*=============================================================================
+
+Library: CppMicroServices
+
+Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+file at the top-level directory of this distribution and at
+https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=============================================================================*/
+
+#include <gtest/gtest.h>
+
+#include <cppmicroservices/Bundle.h>
+#include <cppmicroservices/BundleEvent.h>
+#include <cppmicroservices/BundleContext.h>
+#include "cppmicroservices/Framework.h"
+#include "cppmicroservices/FrameworkEvent.h"
+#include "cppmicroservices/FrameworkFactory.h"
+#include "cppmicroservices/SharedLibraryException.h"
+
+#include "cppmicroservices/servicecomponent/runtime/ServiceComponentRuntime.hpp"
+
+#include "TestInterfaces/Interfaces.hpp"
+#include "TestUtils.hpp"
+
+namespace scr = cppmicroservices::service::component::runtime;
+
+class SharedLibraryExceptionTest : public ::testing::Test
+{
+protected:
+  SharedLibraryExceptionTest()
+    : framework(cppmicroservices::FrameworkFactory().NewFramework())
+  { }
+      
+  virtual ~SharedLibraryExceptionTest() = default;
+
+  virtual void SetUp()
+  {
+    framework.Start();
+    auto context = framework.GetBundleContext();
+
+    auto dsPluginPath = test::GetDSRuntimePluginFilePath();
+    auto dsbundles = context.InstallBundles(dsPluginPath);
+    for (auto& bundle : dsbundles) {
+      bundle.Start();
+    }
+    
+    auto sRef = context.GetServiceReference<scr::ServiceComponentRuntime>();
+    ASSERT_TRUE(sRef);
+    dsRuntimeService = context.GetService<scr::ServiceComponentRuntime>(sRef);
+    ASSERT_TRUE(dsRuntimeService);
+  }
+
+  virtual void TearDown()
+  {
+    framework.Stop();
+    framework.WaitForStop(std::chrono::milliseconds::zero());
+  }
+
+  cppmicroservices::Framework& GetFramework() { return framework; }
+  std::shared_ptr<scr::ServiceComponentRuntime> dsRuntimeService;
+  
+private:
+  cppmicroservices::Framework framework;
+};
+
+TEST_F(SharedLibraryExceptionTest, testDSBundleImmediateTrue)
```

Comment:
In addition to the fake bundles which simulate a shared library loading failure, is it possible to write tests which use mocks to test the behavior of internal DS classes when a `SharedLibraryException` is thrown?

Reply:
Will be looking into using mocks

Reply:
The function that throws due to dlopen/loadlibrary seemed a little difficult to mock.
After reviewing current changes, if the needs are still there, we can create a separate task to add more testpoints using mocks.